### PR TITLE
Add client management

### LINKED ISSUE #142: Add client management
Moved from [Jira](https://jira.osiam.org/browse/OCFJ-11).

Managing OAuth clients is currently not supported by the connector, but seems like a viable feature.

### PR DESCRIPTION
**Pull request by dacrome**
Resolves #142

### CODE REVIEW

**Comment by dacrome** [comment]
Corresponding integration-tests https://github.com/osiam/connector4java-integration-tests/pull/301

File: src/main/java/org/osiam/client/OsiamConnector.java [hunk 1]
```diff
@@ -65,7 +69,8 @@
     private static final int DEFAULT_MAX_CONNECTIONS_PER_ROUTE = 20;
 
     private static final PoolingHttpClientConnectionManager CONNECTION_MANAGER;
-    private static final Client CLIENT;
+    private static final javax.ws.rs.client.Client CLIENT;
```
Comment (wallner) [inline src/main/java/org/osiam/client/OsiamConnector.java]: This is not a traditional constant. Please don't name it all upper case. (We agreed on something similar with the loggers, remember?)

File: src/main/java/org/osiam/client/OsiamConnector.java [hunk 2]
```diff
@@ -65,7 +69,8 @@
     private static final int DEFAULT_MAX_CONNECTIONS_PER_ROUTE = 20;
 
     private static final PoolingHttpClientConnectionManager CONNECTION_MANAGER;
-    private static final Client CLIENT;
+    private static final javax.ws.rs.client.Client CLIENT;
+    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Comment (wallner) [inline src/main/java/org/osiam/client/OsiamConnector.java]: See above, this is not a traditional constant.

File: src/main/java/org/osiam/client/AuthService.java [hunk 3]
```diff
@@ -65,6 +71,7 @@
     private static final String TOKEN_ENDPOINT = "/oauth/token";
     private static final String TOKEN_REVOCATION_ENDPOINT = "/token/revocation";
     private static final String TOKEN_VALIDATION_ENDPOINT = "/token/validation";
+    private static final String CLIENT_ENDPOINT = "/Client";
```
Comment (wallner) [inline src/main/java/org/osiam/client/AuthService.java]: Really a capital 'C'? I'd suggest `/client`

File: src/main/java/org/osiam/client/AuthService.java [hunk 4]
```diff
@@ -333,6 +496,12 @@ private void checkAndHandleResponse(String content, StatusType status, AccessTok
         } else if (status.getStatusCode() == Status.FORBIDDEN.getStatusCode()) {
             String errorMessage = extractErrorMessageForbidden(accessToken);
             throw new ForbiddenException(errorMessage);
+        } else if (status.getStatusCode() == Status.NOT_FOUND.getStatusCode()) {
```
Comment (wallner) [inline src/main/java/org/osiam/client/AuthService.java]: I think we should get rid of this huge else if block. (It's almost a case statement.) using object oriented design and polymorphism could be the way to go.

File: src/main/java/org/osiam/client/AuthService.java (hunk 3 above)
Comment (dacrome) [inline src/main/java/org/osiam/client/AuthService.java]: If you don't like it, please create an issue in https://github.com/osiam/auth-server

File: src/main/java/org/osiam/client/OsiamConnector.java (hunk 1 above)
Comment (dacrome) [inline src/main/java/org/osiam/client/OsiamConnector.java]: done

File: src/main/java/org/osiam/client/OsiamConnector.java (hunk 2 above)
Comment (dacrome) [inline src/main/java/org/osiam/client/OsiamConnector.java]: done

File: src/main/java/org/osiam/client/AuthService.java (hunk 4 above)
Comment (dacrome) [inline src/main/java/org/osiam/client/AuthService.java]: Yes, I also don't like. But shouldn't this be done in an own PR?

**Comment by dacrome** [comment]
I don't like the [current use of the object mapper](https://github.com/osiam/connector4java/pull/172/files#diff-9c26fca76d9cb9c779a35ce8f3b10af5R455) to parse the client object. [This is how it should be done](https://gist.github.com/dacrome/ee576463d6fa8426c393), but I getting an exception which claims that a `MessageBodyWriter` not found for the [Client](https://github.com/dacrome/connector4java/blob/add-client-management/src/main/java/org/osiam/client/oauth/Client.java). Any suggestions?

File: src/main/java/org/osiam/client/AuthService.java (hunk 3 above)
Comment (wallner) [inline src/main/java/org/osiam/client/AuthService.java]: Apparently all SCIM endpoints start with a capital letter too, so please ignore the comment.

File: src/main/java/org/osiam/client/AuthService.java (hunk 4 above)
Comment (wallner) [inline src/main/java/org/osiam/client/AuthService.java]: As I wrote before.

**Comment by wallner** [comment]
@dacrome That's usually an indication that the [jaxrs provider](https://github.com/FasterXML/jackson-jaxrs-providers/) is not on the classpath. Just add them.

**Comment by dacrome** [comment]
> That's usually an indication that the jaxrs provider is not on the classpath. Just add them.

I tried it with `jersey-media-json-jackson` but then the ITs failing because of class not found exceptions because of version conflicts... jersey-jackson dependency hell, I don't like it.

**Comment by wallner** [comment]
My recommendation is to use the [jackson-jaxrs-providers](https://github.com/FasterXML/jackson-jaxrs-providers/) and stay away from jersey for this one. We should probably also shade them away. At least that was @tkrilles idea when I talked to him about it.

**Comment by dacrome** [comment]
I add the dependency

```
<dependency>
  <groupId>com.fasterxml.jackson.jaxrs</groupId>
  <artifactId>jackson-jaxrs-json-provider</artifactId>
  <version>2.5.4</version>
</dependency>
```
but still get the following exception
```
org.osiam.bundled.org.glassfish.jersey.message.internal.MessageBodyProviderNotFoundException: 
MessageBodyWriter not found for media type=application/json,
type=class org.osiam.client.oauth.Client,
genericType=class org.osiam.client.oauth.Client.
```

**Comment by wallner** [comment]
Thank you.
